Add a driver for rkt

achanda opened the pull request:
- Adds a driver for rkt
- Adds associated tests

ryanuber commented:
Awesome! Thanks @achanda. Would it be possible to use the native Go rkt interface [here](https://github.com/coreos/rkt/tree/master/api/v1)? I'm not familiar with it myself but it thinking it will be more resilient than using the CLI.

achanda commented:
@ryanuber ah, I was looking for something like that. Thanks for the pointer. Let me look around.

catsby commented on client/driver/rkt.go:
```diff
@@ -0,0 +1,206 @@
+        }
+        d.logger.Printf("[DEBUG] Started ACI: %q", name)
+        h := &rktHandle{
+                proc:   cmd.Process,
```
Should this be `acmd`? It's using the `cmd` from line 92  above, where we invoke `rkt trust`

achanda replied:
Thanks, fixed.

catsby commented on client/driver/rkt.go:
```diff
@@ -0,0 +1,206 @@
+// and then provide a 5 second grace period before doing a Kill.
+//
+// TODO: allow a 'shutdown_command' that can be executed over a ssh connection
+// to the VM
```
Let's remove this comment, I don't think it applies :smile:

catsby commented:
Thanks for the contribution! Let us know what you find out about the rkt interface Ryan mentioned

achanda commented:
@ryanuber @catsby I had a chat with the rkt guys. While they are working on a way to provide read only info (image etc.), there are no plans to drive rkt through an API. What makes that more difficult is that rkt does not run a daemon in the background. Kubernetes uses rkt in a similar way as this PR. I took a look at their code and noticed that they are spawning the binary in a similar way. However, their rkt package is too tightly integrated with kubernetes.

ryanuber commented on client/driver/rkt.go:
```diff
@@ -0,0 +1,203 @@
+
+        node.Attributes["driver.rkt"] = "1"
+        node.Attributes["driver.rkt.version"] = rktMatches[0]
+        node.Attributes["driver.appc.version"] = appcMatches[1]
```
This should probably still be nested under the `driver.rkt` namespace, since `appc` itself isn't the driver.

ryanuber commented on client/driver/rkt.go:
```diff
@@ -0,0 +1,203 @@
+                return nil, fmt.Errorf("Missing trust prefix for rkt")
+        }
+        trust_args := []string{
+                "sudo",
```
Since we check in the fingerprint that we are running as root, I don't think we will need `sudo` here.

ryanuber commented on client/driver/rkt.go:
```diff
@@ -0,0 +1,203 @@
+                return nil, fmt.Errorf("Missing ACI name for rkt")
+        }
+        run_args := []string{
+                "sudo",
```
Do we need this sudo? (same as above)

ryanuber commented on client/driver/rkt_test.go:
```diff
@@ -0,0 +1,99 @@
+        }
+        if node.Attributes["driver.rkt.version"] == "" {
+                t.Fatalf("Missing Rkt driver version")
+        }
```
Looks like we are missing a test for the appc version from above

ryanuber commented on client/testutil/driver_compatible.go:
```diff
@@ -18,6 +18,12 @@ func QemuCompatible(t *testing.T) {
 }
 
+func RktCompatible(t *testing.T) {
+        if runtime.GOOS != "windows" && syscall.Geteuid() != 0 {
```
This check might be off. Shouldn't we check if we both aren't on windows, and that we are root? So like, `GOOS == windows || Geteuid() != 0` ?

achanda replied:
Ah, you're right. I copied this bit from the functions above it. Both of them are incorrect then. I can fix those in another patch.

ryanuber commented:
@achanda thanks for looking into the client stuff! Understood, and I think this approach is fine if that's the case. I left a few minor comments but overall looking really good! Can you also add some driver documentation in the `website/` directory to detail the config options, limitations, etc?

achanda commented:
@ryanuber all done.

ryanuber commented on client/driver/rkt.go:
```diff
@@ -0,0 +1,201 @@
+                return false, fmt.Errorf("Unable to parse Rkt version string: %#v", rktMatches)
+        }
+
+        node.Attributes["driver.rkt"] = "1"
```
Missed this on the first pass, we set this to `true` in the Docker driver [here](https://github.com/hashicorp/nomad/blob/master/client/driver/docker.go#L68). I think we should keep these uniform.

ryanuber commented on client/driver/rkt_test.go:
```diff
@@ -0,0 +1,102 @@
+func rktLocated() bool {
+        _, err := exec.Command("rkt", "version").CombinedOutput()
+        return err == nil
+}
```
Can we actually combine this with the `RocketCompatible()` function? Seems to be very closely related.

achanda replied:
This just checks if the binary exists. The other one checks if we can run it. You sure those should be combined?

ryanuber replied:
I don't see the value in checking if we are on non-Windows and running as root if we aren't actually going to execute Rocket. Let me know if I'm missing something, though.

ryanuber commented on website/source/docs/drivers/rkt.html.md:
```diff
@@ -0,0 +1,43 @@
+
+The `Rkt` driver will set the following client attributes:
+
+* `driver.rkt` - Set to `1` if Rkt is found on the host node. Nomad determines
```
Change to reflect `true`

ryanuber commented:
Thanks! I took a final pass and just noted a few more (super) minor things. Thanks again, looks like a merge is coming soon!

dadgar commented on client/driver/rkt.go:
```diff
@@ -0,0 +1,201 @@
+        if !ok || trust_prefix == "" {
+                return nil, fmt.Errorf("Missing trust prefix for rkt")
+        }
+        trust_args := []string{
```
I would remove this an inline it into the exec.Command.

dadgar commented on client/driver/rkt.go:
```diff
@@ -0,0 +1,201 @@
+        trust_args := []string{
+                "rkt",
+                "trust",
+                "--prefix=" + trust_prefix,
```
fmt.Sprintf

dadgar commented on client/driver/rkt.go:
```diff
@@ -0,0 +1,201 @@
+        cmd := exec.Command(trust_args[0], trust_args[1:]...)
+        cmd.Stdout = &outBuf
+        cmd.Stderr = &errBuf
+        d.logger.Printf("[DEBUG] Starting rkt command: %q", strings.Join(trust_args, " "))
```
You can then mimic this with cmd.Args

dadgar commented on client/driver/rkt.go:
```diff
@@ -0,0 +1,201 @@
+        if !ok || name == "" {
+                return nil, fmt.Errorf("Missing ACI name for rkt")
+        }
+        run_args := []string{
```
Same here

dadgar commented on client/driver/rkt.go:
```diff
@@ -0,0 +1,201 @@
+        cmd.Stdout = &outBuf
+        cmd.Stderr = &errBuf
+        d.logger.Printf("[DEBUG] Starting rkt command: %q", strings.Join(trust_args, " "))
+        if err := cmd.Start(); err != nil {
```
Start will not wait for the command to exit. You want to use Run

dadgar commented on client/driver/rkt.go:
```diff
@@ -0,0 +1,201 @@
+        }
+        data, err := json.Marshal(pid)
+        if err != nil {
+                log.Printf("[ERR] failed to marshal rkt PID to JSON: %s", err)
```
You will want to pass the logger from the Driver into the Handle. We support redirecting logs to files or other sources so this would break that it interface by using the default logger.

dadgar commented on client/driver/rkt.go:
```diff
@@ -0,0 +1,201 @@
+        }
+        run_args := []string{
+                "rkt",
+                "run", "--interactive", name,
```
I don't think we want to use --interactive. That attaches the pod to `tty`. I think rather we want to use the `status --wait uuid` command: https://coreos.com/rkt/docs/latest/subcommands/status.html

dadgar commented on client/driver/rkt_test.go:
```diff
@@ -0,0 +1,102 @@
+        }
+
+        driverCtx := testDriverContext(task.Name)
+        ctx := testDriverExecContext(task, driverCtx)
```
Need to add this:
defer ctx.AllocDir.Destroy()

dadgar commented on client/driver/rkt_test.go:
```diff
@@ -0,0 +1,102 @@
+                t.Fatalf("Missing appc version for the Rkt driver")
+        }
+}
+
```
Add a test for waiting

achanda commented:
The last patch set should have addressed all of the comments. Thanks guys!

@dadgar is attaching a tty a problem? rkt containers exit on `Ctrl + ]`, I could not figure out how to send that signal with no tty attached. On having a tty attached, they will exit on `Ctrl + C`. Sorry if I am missing anything here.

cbednarski commented:
> is attaching a tty a problem?

Probably, though this depends on the software running inside the container. Some applications behave differently when a TTY is allocated. For example, an application may use interactive password prompts. This will cause the application to hang since Nomad does not run interactively. You can read more about this here https://github.com/mitchellh/packer/issues/1804.

I am not familiar with `rkt`'s behavior in this case but we should be able to send a sigterm or sigkill to a process to terminate it.

achanda commented:
@cbednarski I understand, updated the PR

achanda commented:
I believe I have addressed all comments. This now passes all tests on my box. Please take a look.

ryanuber commented on client/testutil/driver_compatible.go:
```diff
@@ -18,6 +19,15 @@ func QemuCompatible(t *testing.T) {
+        }
+        // else see if rkt exists
+        _, err := exec.Command("rkt", "version").CombinedOutput()
+        return err == nil
```
Shouldn't we skip if rkt isn't installed or can't be executed? There is no test for this bool anywhere.

achanda commented:
@ryanuber Thanks, updated

ryanuber commented on client/testutil/driver_compatible.go:
```diff
@@ -18,6 +19,17 @@ func QemuCompatible(t *testing.T) {
+        }
+        // else see if rkt exists
+        _, err := exec.Command("rkt", "version").CombinedOutput()
+        if err == nil {
```
Looks like this test is inverted

achanda commented:
@ryanuber please let me know if this needs anything else

armon commented on client/driver/rkt.go:
```diff
@@ -0,0 +1,197 @@
+                        "Error running rkt: %s\n\nOutput: %s\n\nError: %s",
+                        err, outBuf.String(), errBuf.String())
+        }
+        d.logger.Printf("[DEBUG] Added trust prefix: %q", trust_prefix)
```
Can you make sure all logging statements have the proper prefix of "driver.rkt"

achanda commented:
@armon updated

ryanuber commented:
@achanda I tried running your tests locally, but I think we have a problem. The `rkt` command that starts the container blocks. In `rkt.go` line 110 (which is part of the `Start()` method), you are calling `Run()` on the command, which blocks until execution finishes. You can use `Start()` on the command to run it and get a handle on it, and track that in the driver. Were you able to get multiple containers to run, or to get the tests to pass without manual intervention? In the current state, it seems that this will always block, causing the tests to eventually time out and fail.

achanda commented:
@ryanuber I changed that to address a review comment. However, I do see your point. The tests do run fine on my box though, I'm not sure how! I started another `rkt` pod manually, independent of nomad, on a box and ran the tests, that went fine. Let me look around more.

achanda commented:
@ryanuber I changed to `Start` and rebased on master. These tests are passing on my box now. Please check.

dadgar commented on client/driver/rkt.go:
```diff
@@ -0,0 +1,197 @@
+        cmd.Stdout = &outBuf
+        cmd.Stderr = &errBuf
+        d.logger.Printf("[DEBUG] driver.rkt: starting rkt command: %q", cmd.Args)
+        if err := cmd.Start(); err != nil {
```
This one should still be Run because you don't want to do rkt run before trust has finished

dadgar commented on client/driver/rkt_test.go:
```diff
@@ -0,0 +1,132 @@
+                Name: "etcd",
+                Config: map[string]string{
+                        "trust_prefix": "coreos.com/etcd",
+                        "name":  "coreos.com/etcd:v2.0.4",
```
You will want to run something that exits to test the wait. This test just times out right now

dadgar commented on website/source/docs/drivers/rkt.html.md:
```diff
@@ -0,0 +1,43 @@
+* `trust_prefix` - **(Required)** The trust prefix to be passed to rkt. Must be reachable from
+the box running the nomad agent.
+* `name` - **(Required)** Fully qualified name of an image to run using rkt
+
```
Add an exec option to override the command to run in the image.

dadgar commented:
Can you also run gofmt on all of the files. They are off a little. You can do this by running `make format`

ryanuber commented:
@achanda thanks for working through this with us! We are amped to have rkt supported in Nomad.

achanda commented:
Thanks @ryanuber and all other reviewers :)

dadgar commented:
@achanda: This was missed in the original review but the driver does not do resource enforcement which is part of the driver API. This poses a problem because Nomad can make no guarantee about the cpu/memory usage of the task and which ports it is using. Given your experience with `rkt` do you have any thoughts on improving this driver. Otherwise we may have to remove it until it can be implemented in a fashion that respects the resource constraints.

achanda commented:
@dadgar CPU, memory and network isolation can be definitely be achieved using `rkt`. I will start working on those.

dadgar commented:
@achanda: Awesome look forward to it. This was caught as we were reviewing for the Nomad 0.2 release. So to make the cut, this would need to be done by Monday, otherwise we could just disable for 0.2 and enable it in a latter version. Sorry for the short notice.

achanda commented:
@dadgar ok, let me see what I can do.

achanda commented:
@dadgar here is what I found, since the `rkt` CLI does not accept CPU and memory restrictions directly, the easiest way to achieve this will be using `systemd-run`, which will add another dependency. I am not familiar with the `--pod-manifest` flag, that can work but will need more effort than I can manage by Monday. Another way of doing this will be to add these options to the `rkt` CLI itself, but that won't work in a few days. Please let me know if we want to run `rkt` via systemd, otherwise the only option might be to remove this driver for now.

dadgar commented:
@achanda: The system-d run solution is not quite enough. That provides CPU/Memory isolation but it doesn't fix port allocations. Whatever is in the manifest is what will be used. That needs to be fixed as well. My opinion is these really should be adjustable during `rkt run` similar to docker. 

So we discussed it and we are going to leave the driver in for 0.2 but update the documentation to mark it as experimental. This allows the driver to be improved as changes to the rkt ecosystem occur.
